php/php-rr-sdk.md

```diff
@@ -0,0 +1,692 @@
+
+if ($t->isFired()) {
+    // ...
+}
+```
+
+### Queries
+Workflow queries can be easily implemented using dedicated query method of workflow. The state of the workflow can be 
+extracted from the properties of the object.
+
+```php
+use Temporal\Workflow;
+
+class DemoWorkflow extends Workflow\Workflow
+{
+    private $step = 0;
+
+    public function queryStep(): int 
+    {
+        return $this->step;
+    }
+
+    public function run(string $input): string
+    {
+        yield $this->activities->doSomething($input);
+        $this->step++;
+
+        // ...
+    }
+}
+```
+
+> Query methods must be registered explicitly.
+
+An alternative approach can be based on registering the callback function during the workflow execution:
+
+```php
+yield new QueryHandler('something', function() use (&$step) {
+    return $step;
+});
+```
+
+> Such an approach is more suitable for dynamic workflows. The principle of work is identical to pre-registered queries.
+
+#### Using Annotations
+For more advanced frameworks it should be possible to register query method using annotation or attribute (PHP8) as 
+in Java SDK.
+
+```php
+/**
+ * @Workflow\QueryMethod(name="something")
+ */
+public function querySomething(string $world): string 
+{
+    return 'hello';
+}
+```
+
+### Signals
+Unlike queries the signals are fist-class citizen of actual workflow code. Unlike activities the signal can be triggered
+at any moment, without prior expectation by workflow. 
+
+Signal method can be defined explicitly:
+
+```php
+class UploadWorkflow extends Workflow\Workflow
+{
+    /** @Temporal\SignalMethod(name="signal") */
+    public function handleSignal($input):void
+    {
+    
+    }
+
+    public function run(string $input): string
+    {
+        
+    }
+}
+```
+
+Or at runtime:
+
+```php
+public function run(string $input): string
+{
+    yield new SignalHandler('name', function($input){
+    
+    });       
+}
+```
+
+> Need Temporal authors feedback.
+
+#### Handle Signals
+Incoming signals can be handled during any workflow `yield` call. Workflow should allow to explicitly wait for signal:
+
+```php
+public function run(string $input): string
+{
+    yield new SignalHandler('name', function($input){
+        // do something
+    });       
+
+    yield new WaitSignal('name');
+}
+```
+
+### Sessions
+Sessions can be created as sub-activity group, using similar approach as `withTimeout` and other methods. However,
+the explicit yield required. 
+
+```php
+public function run(string $input): string
+{
+    $session = yield $this->activities->createSession(100, 100); // timeouts
+    
+    $result = yield $session->doSomething($input)->get();
+    $result = yield $session->somethingElse($result)->get();
+    
+    yield $session->complete();
+    
+    return $result;
+}
+```
+
+### Deterministic Time
+Workflows must avoid calling SPL functions `time()` and `date()`. Context method must be used instead:
+
+```php
+$ctx->getNow(); //DateTimeImmutable object.
```

Comment:
why not ctx->time() and ctx->date()?

Reply:
The time() function in PHP returns UNIX timestamp (int), date() returns formatted date. None of them are useful for the time manipulation. 

The go-to solution is to use `DateTime` interfaces which allow safe date/time transformations. So `getNow()` or `now()` are less ambiguous for developers since it does not create a bridge for native functions. However, we can add them as well without any issue, it could be:

```php
$ctx->now(); // date time
$ctx->time(); // Unix timestamp
$ctx->date('D-Y-m') // formatted date

// last two equals to 
$ctx->now()->getTimestamp();
$ctx->now()->format('D-Y-m');
```